src/terminal_chat_console.cpp

```diff
@@ -0,0 +1,423 @@
+			// delete word to the right
+			m_chat_backend.getPrompt().cursorOperation(
+				ChatPrompt::CURSOROP_DELETE,
+				ChatPrompt::CURSOROP_DIR_RIGHT,
+				ChatPrompt::CURSOROP_SCOPE_WORD);
+			break;
+		case 21:
+			// Ctrl-U pressed
+			// kill line to left end
+			m_chat_backend.getPrompt().cursorOperation(
+				ChatPrompt::CURSOROP_DELETE,
+				ChatPrompt::CURSOROP_DIR_LEFT,
+				ChatPrompt::CURSOROP_SCOPE_LINE);
+			break;
+		case 11:
+			// Ctrl-K pressed
+			// kill line to right end
+			m_chat_backend.getPrompt().cursorOperation(
+				ChatPrompt::CURSOROP_DELETE,
+				ChatPrompt::CURSOROP_DIR_RIGHT,
+				ChatPrompt::CURSOROP_SCOPE_LINE);
+			break;
+		case KEY_TAB:
+			// Tab pressed
+			// Nick completion
+			m_chat_backend.getPrompt().nickCompletion(m_nicks, false);
+			break;
+		default:
+			// Add character to the prompt,
+			// assuming UTF-8.
+			if (IS_UTF8_MULTB_START(ch)) {
+				m_pending_utf8_bytes.append(1, (char)ch);
+				m_utf8_bytes_to_wait += UTF8_MULTB_START_LEN(ch) - 1;
+			} else if (m_utf8_bytes_to_wait != 0) {
+				m_pending_utf8_bytes.append(1, (char)ch);
+				m_utf8_bytes_to_wait--;
+				if (m_utf8_bytes_to_wait == 0) {
+					std::wstring w = utf8_to_wide(m_pending_utf8_bytes);
+					m_pending_utf8_bytes = "";
+					// hopefully only one char in the wstring...
+					for (size_t i = 0; i < w.size(); i++) {
+						m_chat_backend.getPrompt().input(w.c_str()[i]);
+					}
+				}
+			} else if (IS_ASCII_PRINTABLE_CHAR(ch)) {
+				m_chat_backend.getPrompt().input(ch);
+			} else {
+				// Silently ignore characters we don't handle
+
+				//warningstream << "Pressed invalid character '"
+				//	<< keyname(ch) << "' (code " << itos(ch) << ")" << std::endl;
+			}
+			break;
+	}
+}
+
+void TerminalChatConsole::step(int ch)
+{
+	bool complete_redraw_needed = false;
+
+	// empty queues
+	while (!m_chat_interface->outgoing_queue.empty()) {
+		ChatEvent evt = m_chat_interface->outgoing_queue.pop_frontNoEx();
+		switch (evt.type) {
+			case CET_NICK_REMOVE:
+				m_nicks.remove(evt.nick);
+				break;
+			case CET_NICK_ADD:
+				m_nicks.push_back(evt.nick);
+				break;
+			case CET_CHAT:
+				complete_redraw_needed = true;
+				// This is only used for direct replies from commands
+				// or for lua's print() functionality
+				m_chat_backend.addMessage(L"", evt.evt_msg);
+				break;
+		};
+	}
+	while (!m_log_output.queue.empty()) {
+		complete_redraw_needed = true;
+		std::pair<LogLevel, std::string> p = m_log_output.queue.pop_frontNoEx();
+		if (p.first > m_log_level)
+			continue;
+
+		m_chat_backend.addMessage(
+			utf8_to_wide(Logger::getLevelLabel(p.first)),
+			utf8_to_wide(p.second));
+	}
+
+	// handle input
+	if (!m_esc_mode) {
+		handleInput(ch, complete_redraw_needed);
+	} else {
+		switch (ch) {
+			case ERR: // no input
+				break;
+			case 27: // ESC
+				// Toggle ESC mode
+				m_esc_mode = !m_esc_mode;
+				break;
+			case 'L':
+				m_log_level--;
+				m_log_level = MYMAX(m_log_level, 1); // LL_NONE isn't accessible
```

Comment:
This should use `(int)LL_ERROR` instead of hard coding it's current integer equivalent.